language/vm/vm_runtime/vm_runtime_types/src/type_context.rs

```diff
@@ -0,0 +1,40 @@
+use crate::loaded_data::{struct_def::StructDef, types::Type};
+use vm::errors::VMInvariantViolation;
+
+pub struct TypeContext(Vec<Type>);
+
+impl TypeContext {
+    pub fn new(ty: Vec<Type>) -> Self {
+        Self(ty)
+    }
+
+    pub fn id(num_type_args: u16) -> Self {
+        Self((0..num_type_args).map(Type::TypeVariable).collect())
+    }
+
+    pub fn subst_type(&self, ty: &Type) -> Result<Type, VMInvariantViolation> {
+        Ok(match ty {
+            Type::TypeVariable(idx) => self.get_type(*idx)?,
+            Type::Reference(ty) => Type::Reference(Box::new(self.subst_type(ty)?)),
+            Type::MutableReference(ty) => Type::MutableReference(Box::new(self.subst_type(ty)?)),
+            Type::Struct(s) => Type::Struct(self.subst_struct_def(s)?),
+            id => id.clone(),
+        })
+    }
+
+    pub fn subst_struct_def(&self, def: &StructDef) -> Result<StructDef, VMInvariantViolation> {
+        Ok(StructDef::new(
+            def.field_definitions()
+                .iter()
+                .map(|ty| self.subst_type(ty))
+                .collect::<Result<Vec<Type>, VMInvariantViolation>>()?,
```

Comment:
`collect::<Result<Vec<_>, _>>` or `collect::<Result<_, _>>` may be enough, but it's personal taste.